plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetadataFileOperations.java

```diff
@@ -334,6 +334,34 @@ public void testShowStatsForTableWithFilter()
                         .build());
     }
 
+    @Test
+    public void testPredicateWithVarcharCastToDate()
+    {
+        assertUpdate("CREATE TABLE test_varchar_as_date_predicate(a varchar) WITH (partitioning=ARRAY['truncate(a, 4)'])");
+        assertUpdate("INSERT INTO test_varchar_as_date_predicate VALUES '2001-01-31'", 1);
+        assertUpdate("INSERT INTO test_varchar_as_date_predicate VALUES '2005-09-10'", 1);
+
+        assertFileSystemAccesses("SELECT * FROM test_varchar_as_date_predicate",
+                ImmutableMultiset.builder()
+                        .addCopies(new FileOperation(MANIFEST, INPUT_FILE_GET_LENGTH), 4)
+                        .addCopies(new FileOperation(MANIFEST, INPUT_FILE_NEW_STREAM), 4)
+                        .addCopies(new FileOperation(METADATA_JSON, INPUT_FILE_NEW_STREAM), 1)
+                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_GET_LENGTH), 1)
+                        .addCopies(new FileOperation(SNAPSHOT, INPUT_FILE_NEW_STREAM), 1)
+                        .build());
+
+        assertFileSystemAccesses("SELECT * FROM test_varchar_as_date_predicate WHERE CAST(a AS date) >= DATE '2005-01-01'",
```

Comment:
I find this addition hidden gem.
Could we get this documented ?

We should probably add here in the test hints that the casting improvement is not for everything.

e.g. : `SELECT * FROM test_varchar_as_date_predicate WHERE year(CAST(a AS date)) >= 2005` does not benefit anymore of the the improvement.

Reply:
> I find this addition hidden gem.
> Could we get this documented ?

"Trino does as much optimizations as possible".
I think we should not try to document all the optimizations. What for?

> e.g. : `SELECT * FROM test_varchar_as_date_predicate WHERE year(CAST(a AS date)) >= 2005` does not benefit anymore of the the improvement.

yes, not yet.